### DISCUSSION TITLE
ch4/ch4-01

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#188] [General]
# ch4/ch4-01

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch4/ch4-01.html

**Comment by lingengyuan** [comment]
练习4.2：
package main

import (
	"crypto/sha256"
	"crypto/sha512"
	"fmt"
	"os"
)

func main() {
	var flag string
	var str = os.Args[1]
	if len(os.Args) > 1 {
		flag = os.Args[2]
	} else {
		flag = "256"
	}
	if flag == "256" {
		str256 := sha256.Sum256([]byte(str))
		fmt.Printf("%x\n", str256)
	} else if flag == "384" {
		str384 := sha512.Sum384([]byte(str))
		fmt.Printf("%x\n", str384)
	} else if flag == "512" {
		str512 := sha512.Sum512([]byte(str))
		fmt.Printf("%x\n", str512)
	}
}

**Comment by lingengyuan** [comment]
练习4.1：

func PopCount(x, y string) (num int) {
	X := sha256.Sum256([]byte(x))
	Y := sha256.Sum256([]byte(y))
	for i := 0; i < len(X); i++ {
		for m := 1; m <= 8; m++ {
			if (X[i] >> uint(m)) != (Y[i] >> uint(m)) {
				num++
			}
		}
	}
	return num
}

**Reply by xianren68** [reply]
这样比较每一位是否相同应该有问题吧

**Comment by xianren68** [comment]
### 4.1
```
package main

import (
	"crypto/sha256"
	"fmt"
)

func main() {
	c1 := sha256.Sum256([]byte("x"))
	c2 := sha256.Sum256([]byte("X"))
	fmt.Println(differentCount(c1, c2))
}

func differentCount(c1, c2 [32]uint8) (count int) {
	count = 0
	for i, _ := range c1 {
		count += popCount(c1[i], c2[i])
	}
	return
}

// 比较每一个bit位是否相同
func popCount(x, y byte) int {
	count := 0
	for i := 0; i < 8; i++ {
		xb := x % 2
		yb := y % 2
		if xb != yb {
			count++
		}
		x = x >> 1
		y = y >> 1
	}
	return count
}
```

**Comment by xianren68** [comment]
### 4.2
```
func main() {
	var flag string
	if len(os.Args) > 1 {
		flag = os.Args[1]
	}
	// 标准输入的字符串
	var input string
	fmt.Scanln(&input)
	if flag == "" {
		fmt.Println(sha256.Sum256([]byte(input)))
	} else if flag == "SHA384" {
		fmt.Println(sha512.Sum384([]byte(input)))
	} else if flag == "SHA512" {
		fmt.Println(sha512.Sum512([]byte(input)))
	} else {
		fmt.Println("请输入正确的加密方式")
	}

}
```

**Comment by zb81** [comment]
4-2：
```zsh
go run main.go --method=1
Enter a string: abcd
88d4266fd4e6338d13b845fcf289579d209c897823b9217da3e161936f031589

```

```go
package main

import (
	"crypto/sha256"
	"crypto/sha512"
	"flag"
	"fmt"
	"log"
)

func main() {

	var method string

	flag.StringVar(&method, "method", "1", "编码方式（1 - SHA256, 2 - SHA384，3 - SHA512）")
	flag.Parse()

	ValidateMethod(method)

	fmt.Print("Enter a string: ")
	var input string
	fmt.Scanln(&input)
	var inputBytes = []byte(input)
	if method == "1" {
		fmt.Printf("%x\n", sha256.Sum256(inputBytes))
	} else if method == "2" {
		fmt.Printf("%x\n", sha512.Sum384(inputBytes))
	} else {
		fmt.Printf("%x\n", sha512.Sum512(inputBytes))
	}
}

func ValidateMethod(m string) {
	if m != "1" && m != "2" && m != "3" {
		log.Fatalf("method %s is not supported.", m)
	}
}
```

**Comment by ReReniao** [comment]
练习4.1 

```go
package main

import (
	"crypto/sha256"
	"fmt"
)

// 每个8位有几个位为1的对照表
var pc [256]byte

// 初始化pc byte数组
func init() {
	for i := range pc {
		pc[i] = pc[i/2] + byte(i&1)
	}
}

func main() {
	c1 := sha256.Sum256([]byte("X"))
    c2 := sha256.Sum256([]byte("X"))
	fmt.Printf("%d",compareSha256(c1, c2))
}

// 返回有几个1位
func Popcount(x byte) int {
	return int(pc[x])
}


func compareSha256(x, y [32]byte) int{
	count := 0
	for i := 0; i < 8; i++ {  // 每8位对比一次
		count += Popcount((x[i]^y[i]))
	}
	return count
}
```

**Comment by SuperZ3** [comment]
练习 4.1
思路，异或后求 1 的个数
```go
package main

import (
	"crypto/sha256"
	"fmt"
)

func popCount(num uint8) int {
	result := 0
	for num > 0 {
		num &= (num - 1)
		result++
	}
	return result
}

func countDiff(a, b [32]uint8) int {
	result := 0
	for i, aValue := range a {
		bValue := b[i]
		diffValue := aValue ^ bValue
		result += popCount(diffValue)
	}
	return result
}

func main() {
	c1 := sha256.Sum256([]byte("x"))
	c2 := sha256.Sum256([]byte("X"))
	diff := countDiff(c1, c2)
	fmt.Println(diff)
}
```
4.2 没看懂题目，额，，，

**Comment by lyb88999** [comment]
func HashPopCount(SHA1 [32]uint8, SHA2 [32]uint8) int {
	    var same [32]uint8
	    for i := 0; i < 32; i++ {
		    // 异或相同为0 不同为1 只需要统计same数组中0的个数就是SHA1和SHA2相同位的个数
		    same[i] = SHA1[i] ^ SHA2[i]
	    }
	    sum := 0
	    for i := 0; i < 32; i++ {
		    for same[i] != 0 {
			    if same[i]&1 == 0 {
				    sum += 1
			    }
			    same[i] >>= 1
		    }
	    }
	    return sum
    }

**Comment by hongjiala** [comment]
package main

import (
	"crypto/sha256"
	"fmt"
)

func main() {
	c1 := sha256.Sum256([]byte("x"))
	c2 := sha256.Sum256([]byte("X"))

	count := countDiffBits(c1, c2)
	fmt.Printf("%x\n%x\n%t\n%[1]T\n", c1, c2, c1 == c2)
	fmt.Println(count)
	// Output:
	// 2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881
	// 4b68ab3847feda7d6c62c1fbcbeebfa35eab7351ed5e78f4ddadea5df64b8015
	// false
	// [32]uint8
}

func countDiffBits(c1, c2 [32]byte) int {
	count := 0
	for i := 0; i < 32; i++ {
		for j := 0; j < 8; j++ {
			if c1[i]&(1<<uint(j)) != c2[i]&(1<<uint(j)) {
				count++
			}
		}
	}
	return count
}

**Comment by hongjiala** [comment]
package main

import (
	"crypto/sha256"
	"crypto/sha512"
	"fmt"
	"os"
)

func main() {
	var flag string
	var str = os.Args[0]
	if len(os.Args) > 1 {
		flag = os.Args[1]
	} else {
		flag = "256"
	}
	switch flag {
	case "256":
		fmt.Printf("%x\n", sha256.Sum256([]byte(str)))
	case "384":
		fmt.Printf("%x\n", sha512.Sum384([]byte(str)))
	case "512":
		fmt.Printf("%x\n", sha512.Sum512([]byte(str)))
	default:
		fmt.Printf("%x\n", sha256.Sum256([]byte(str)))
	}
}

**Comment by xubai-owo** [comment]
4.1
```go
func diffSha256(x, y [32]byte) int {
	count := 0
	for i := range x {
		count += int(pc[x[i]^y[i]])
	}
	return count
}
```

**Reply by w2807** [reply]
4.1
```go
package main

import (
	"crypto/sha256"
	"fmt"
)

func main() {
	c1 := sha256.Sum256([]byte("hello world"))
	c2 := sha256.Sum256([]byte("Sayonara"))
	fmt.Println(popCount(c1, c2))
}

func popCount(c1, c2 [32]byte) int {
	cnt := 0
	for i := range c1 {
		x := c1[i] ^ c2[i]
		for x != 0 {
			cnt++
			x &= x - 1
		}
	}
	return cnt
}
```

**Comment by pyhita** [comment]
```go
type Algo string

const (
	SHA256 Algo = "sha256"
	SHA384 Algo = "sha384"
	SHA512 Algo = "sha512"
)

func main() {
	algo := flag.String("algo", "sha256", "specify the hash algorithm")
	flag.Parse()

	// receive input
	var s string
	input := bufio.NewScanner(os.Stdin)
	if input.Scan() {
		s = input.Text()
		fmt.Println("Input:", s)
	}

	switch Algo(*algo) {
	case SHA256:
		c1 := sha256.Sum256([]byte(s))
		fmt.Printf("%x\n", c1)
	case SHA384:
		c1 := sha512.Sum384([]byte(s))
		fmt.Printf("%x\n", c1)
	case SHA512:
		c1 := sha512.Sum512([]byte(s))
		fmt.Printf("%x\n", c1)
	default:
		fmt.Errorf("unknown hash algorithm: %s", *algo)
		os.Exit(1)
	}
}
```

**Reply by Raoxu20030404** [reply]
package main

import (
	"crypto/sha256"
	"fmt"
)

var pc [256]byte

// init 用来做“包级初始化”，在 main() 之前自动执行
func init() {
	for i := range pc {
		// 一个数的pc = 去掉最后位的pc + 最低位的值
		pc[i] = pc[i>>1] + byte(i&1)
	}
}
func main() {
	c1 := sha256.Sum256([]byte("x"))
	c2 := sha256.Sum256([]byte("X"))

	//fmt.Printf("c1 = %x\n", c1)
	//fmt.Printf("c2 = %x\n", c2)

	diff := compareBit(c1, c2)
	fmt.Println(diff) // 125
}

// 比较两个数据有多少不同的位
func compareBit(c1, c2 [32]byte) int {
	ret := 0

	length := len(c1)

	for i := 0; i < length; i++ {
		diff := c1[i] ^ c2[i] // 记录两个数据相同位置不同处记为1
		ret += popCount(diff)
	}

	return ret
}

// 计算该数据的二进制形式下有多少位是1
func popCount(x byte) int {
	count := 0
	for x != 0 {
		x &= x - 1
		count++
	}

	return count
}

**Reply by Raoxu20030404** [reply]
package main

import (
	"crypto/sha256"
	"crypto/sha512"
	"flag"
	"fmt"
)

var sha *int = flag.Int("sha", 256, "select 256 or 512")

func main() {
	/*
		go run 03_sha.go 				默认SHA256
		go run 03_sha.go -sha 256		选择SHA256
		go run 03_sha.go -sha 384		选择SHA384
		go run 03_sha.go -sha 512		选择SHA512
	*/
	// 解析 flag
	flag.Parse()

	// 读取命令行参数 flag.Args避免读入flag
	for _, filename := range flag.Args() {
		fileHash := generateSHA([]byte(filename), *sha)
		fmt.Printf("filename  = %s\tsha = %x\n", filename, fileHash)
	}

	/* 输出结果
	$  go run 03_sha.go -sha 512 filename1 filename2 filename3
	filename  = filename1   sha = c3294af2171c2be30bb9ebccdf8afb5812d031bfc2f6d1091f5d8f63fbd2e86912a4f0cdee929c86eb9fb5fae6a2f4a0a21dee2d16da2707ddef99d531552641
	filename  = filename2   sha = df41223cf3fdaa652376440c9dadd53607008eec51a70da1a5f6fbd193d2d3e7eac74b65e7c7b6bde0a53b2a6b42a4a6f6e7f833cdc41b1c6640724a4a4753af
	filename  = filename3   sha = 4b1bcd6808a1f5a45d024f14e823891f16ced63e5f447d0e37a00e2de3f1139a4d11f9d398a56621229bf5f9cb30615ddaddb36ca96c9f6755c22fac0a0ac5a6
	*/
}

func generateSHA(aim []byte, s int) interface{} {
	switch s {
	case 384:
		return sha512.Sum384(aim)
	case 512:
		return sha512.Sum512(aim)
	default:
		return sha256.Sum256(aim)
	}
}
